packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4157,3 +4164,261 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
   @override
   bool get isActionEnabled => state._value.selection.isValid && !state._value.selection.isCollapsed;
 }
+
+/// A void function that takes a [TextEditingValue].
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and calls [onTriggered] when a
+/// shortcut is triggered that would affect the state of the [controller].
+class _TextEditingHistory extends StatefulWidget {
+  /// Creates an instance of [_TextEditingHistory].
+  const _TextEditingHistory({
+    Key? key,
+    required this.child,
+    required this.controller,
+    required this.onTriggered,
+  }) : super(key: key);
+
+  /// The child widget of [_TextEditingHistory].
+  final Widget child;
+
+  /// The [TextEditingController] to save the state of over time.
+  final TextEditingController controller;
+
+  /// Called when an undo or redo is received.
+  ///
+  /// If the state would still be the same before and after the undo/redo, this
+  /// will still be called. For example, receiving a redo when there is nothing
+  /// to redo will call this method with no change to the controller.
+  ///
+  /// It is NOT called when the controller is changed for reasons other than
+  /// undo/redo.
+  final TextEditingValueCallback onTriggered;
```

Comment:
continue on the called with the same state.

Isn't the `_update` method short-circuited if the state is the same (line 4233)?

Reply:
Updated the docs.